fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+                )
+
+        if init_state_local_model:
+            local_model.train()
+
+        return torch.cat(local_buffer, dim=0), torch.cat(init_global_buffer, dim=0)
+
+    def predict(self, input: torch.Tensor) -> Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]:
+        """
+        Computes the predictions for both the GLOBAL and LOCAL models and pack them into the prediction dictionary
+
+        Args:
+            input (torch.Tensor): Inputs to be fed into both models.
+
+        Returns:
+            Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]: A tuple in which the first element
+            contains predictions indexed by name and the second element contains intermediate activations
+            index by name.
+
+        Raises:
+            ValueError: Occurs when something other than a tensor or dict of tensors is returned by the model
+            forward.
+        """
+        if isinstance(self.model, MoonModel):
+            global_preds, _ = self.global_model(input)
+            local_preds, features = self.model(input)
+        else:
+            global_preds = self.global_model(input)
+            local_preds = self.model(input)
+            features = {}
+
+        # Here we assume that global and local preds are simply tensors
+        # TODO: Perhaps loosen this at a later date.
+        if not isinstance(global_preds, torch.Tensor):
+            global_preds = global_preds["prediction"]
+        if not isinstance(local_preds, torch.Tensor):
+            local_preds = local_preds["prediction"]
+
+        assert isinstance(global_preds, torch.Tensor)
+        assert isinstance(local_preds, torch.Tensor)
+
+        if self.mkmmd_loss_weight != 0:
+            if not isinstance(self.model, MoonModel) or not isinstance(self.init_global_model, MoonModel):
+                AssertionError(
+                    "To compute the MK-MMD loss, the client model and the init_global_model must be of type MoonModel."
+                )
+            _, init_global_features = self.init_global_model(input)
+            features.update({"init_global_features": init_global_features["features"]})
+
+        return {"global": global_preds, "local": local_preds}, features
+
+    def compute_loss_and_additional_losses(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+        """
+        Computes the loss and any additional losses given predictions of the model and ground truth data.
+
+        Args:
+            preds (Dict[str, torch.Tensor]): Prediction(s) of the model(s) indexed by name.
+            features (Dict[str, torch.Tensor]): Feature(s) of the model(s) indexed by name.
+            target (torch.Tensor): Ground truth data to evaluate predictions against.
+
+        Returns:
+            Tuple[torch.Tensor, Union[Dict[str, torch.Tensor], None]]; A tuple with:
```

Comment:
This specifies the type for the Dict as being `Union[Dict[str, torch.Tensor], None]` but in the method signature, None isn't an option.